File: C#/Leetcode/Array/ProductofArrayExceptSelf.cs
using System;

namespace LeetcodeSolutions.Array
{
    // Leetcode 238
    // Input:  [1,2,3,4]
    // Output: [24,12,8,6]
    class ProductofArrayExceptSelf
    {
        // Runtime : 472 ms
        // Tx = O(n) {O(2n), n: length of nums} 
        // Sx = O(1)
        public int[] ProductExceptSelf(int[] nums)
        {
            if (nums == null || nums.Length == 0)
                return nums;

            int length = nums.Length;
            int[] output = new int[length];
            output[0] = 1;

            int productSoFar = nums[0];

            for (int index = 1; index < length; index++)
            {
                output[index] = productSoFar;
                productSoFar *= nums[index];
            }

            productSoFar = nums[length - 1];
            for (int index = length - 2; index >= 0; index--)
            {
                output[index] *= productSoFar;
                productSoFar *= nums[index];
            }

            return output;
        }
    }
}

File: Java/PrefixSum/ProductofArrayExceptSelf.java
package PrefixSum;

public class ProductofArrayExceptSelf {
    // Leetcode 238: https://leetcode.com/problems/product-of-array-except-self/description/
    // Submission Detail: https://leetcode.com/problems/product-of-array-except-self/submissions/1670802006

    //Input: nums = [1,2,3,4]
    //Output: [24,12,8,6]

    //Input: nums = [-1,1,0,-3,3]
    //Output: [0,0,9,0,0]

    public int[] productExceptSelf(int[] nums) {
        int length = nums.length, productSoFar = nums[0];
        int[] result = new int[length];
        result[0] = 1;

        // Step 1: Store the multipled values from left to right at all indexes except first
        for(int index=1; index<length; index++) {
            result[index] = productSoFar;
            productSoFar *= nums[index];
        }

        // Step 2: Store the multipled values from left to right at all indexes except last
        productSoFar = nums[length-1];
        for(int index=length-2; index>=0; index--) {
            result[index] *= productSoFar;
            productSoFar *= nums[index];
        }

        return result;
    }
}
